### PR TITLE
feat: add "banana in a box" rule

### PR DESCRIPTION
**Pull request by wKoza**
Part of #314

### CODE REVIEW

**Comment by mgechev** [comment]
Haha, I like the error messages. Would be awesome if we're able to introduce this rule as part of the next release as well!

File: src/bananaInBoxRule.ts [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+
+const BoxInABananaOpen = '([';
+const BoxInABananaClose = '])';
+const BananaInABoxOpen = '[(';
+const BananaInABoxClose = ')]';
+const BoxInABananaRe = new RegExp('\\(\\[(.*?)\\]\\)(.*?)');
+
+const getReplacements = (text: ast.BoundEventAst, absolutePosition: number) => {
+  const expr: string = (text.sourceSpan as any).toString();
+  const internalStart = expr.indexOf('([');
```
Comment (mgechev) [inline src/bananaInBoxRule.ts]: Let's use the constants defined above.

File: src/bananaInBoxRule.ts [hunk 2]
```diff
@@ -0,0 +1,76 @@
+import * as Lint from 'tslint';
+import * as ts from 'typescript';
+import * as ast from '@angular/compiler';
+import {BasicTemplateAstVisitor} from './angular/templates/basicTemplateAstVisitor';
+import {NgWalker} from './angular/ngWalker';
+
+
+const BoxInABananaOpen = '([';
```
Comment (mgechev) [inline src/bananaInBoxRule.ts]: Lets rename it to `ValidSyntaxOpen` or something similar. I'm getting confused reading `BoxInABanana*`, wondering if that was the valid or invalid syntax.

File: src/bananaInBoxRule.ts [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+  public apply(sourceFile:ts.SourceFile): Lint.RuleFailure[] {
+
+    return this.applyWithWalker(
+      new NgWalker(sourceFile,
+        this.getOptions(), {
+          templateVisitorCtrl: BananaInBoxTemplateVisitor,
+        }));
+  }
+}
```
Comment (mgechev) [inline src/bananaInBoxRule.ts]: Newline.

File: src/bananaInBoxRule.ts [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+class BananaInBoxTemplateVisitor extends BasicTemplateAstVisitor {
+
+  visitEvent(prop: ast.BoundEventAst, context: any): any {
+
+    if (prop.sourceSpan) {
+      // Note that will not be reliable for different interpolation symbols
+      let error = null;
+      const expr: any = (<any>prop.sourceSpan).toString();
+        if (BoxInABananaRe.test(expr)) {
+        error = 'The box is in a banana! Expecting Banana in a box [(expr)]';
```
Comment (mgechev) [inline src/bananaInBoxRule.ts]: This warning message is fun but lets change it to something like: "Invalid binding syntax. Use [(expr)] instead".

Comment (wKoza) [inline src/bananaInBoxRule.ts]: No problem ;) . Done

File: src/bananaInBoxRule.ts (hunk 3 above)
Comment (wKoza) [inline src/bananaInBoxRule.ts]: Done

File: src/bananaInBoxRule.ts (hunk 2 above)
Comment (wKoza) [inline src/bananaInBoxRule.ts]: Sure. Done

**Comment by mgechev** [comment]
Thanks! Will take a look tomorrow & merge/comment.

**Pull request merged**